File: java/backend/src/main/java/com/etour/app/invoice/controller/InvoiceController.java
package com.etour.app.invoice.controller;

import org.springframework.http.HttpHeaders;
import org.springframework.http.MediaType;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.GetMapping;
import org.springframework.web.bind.annotation.PathVariable;
import org.springframework.web.bind.annotation.PostMapping;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.RestController;

import com.etour.app.email.service.EmailService;
import com.etour.app.email.util.EmailTemplateUtil;
import com.etour.app.invoice.dto.InvoiceResponseDTO;
import com.etour.app.invoice.service.InvoiceService;
import com.etour.app.invoice.util.InvoicePdfGenerator;

@RestController
@RequestMapping("/api/invoices")
public class InvoiceController {

    private final InvoiceService invoiceService;
    private final EmailService emailService;

    public InvoiceController(
            InvoiceService invoiceService,
            EmailService emailService
    ) {
        this.invoiceService = invoiceService;
        this.emailService = emailService;
    }

    @GetMapping("/{bookingId}/download")
    public ResponseEntity<byte[]> downloadInvoice(@PathVariable Long bookingId) {

        // 1. Generate invoice data
        InvoiceResponseDTO invoice = invoiceService.generateInvoice(bookingId);

        // 2. Generate PDF
        byte[] pdfBytes = InvoicePdfGenerator.generateInvoicePdf(invoice);

        // 3. Prepare headers
        HttpHeaders headers = new HttpHeaders();
        headers.setContentType(MediaType.APPLICATION_PDF);
        headers.setContentDispositionFormData(
                "attachment",
                invoice.getInvoiceNumber() + ".pdf"
        );

        // 4. Return response
        return ResponseEntity
                .ok()
                .headers(headers)
                .body(pdfBytes);
    }
    
    /**
     * Manual resend invoice email
     * Used by support/admin or customer retry
     */
    @PostMapping("/{bookingId}/resend-email")
    public ResponseEntity<String> resendInvoiceEmail(
            @PathVariable Long bookingId
    ) {

        // 1️⃣ Generate invoice (validates CONFIRMED + SUCCESS internally)
        InvoiceResponseDTO invoice =
                invoiceService.generateInvoice(bookingId);

        // 2️⃣ Generate PDF
        byte[] pdfBytes =
                InvoicePdfGenerator.generateInvoicePdf(invoice);

        // 3️⃣ Send email (async, non-blocking)
        emailService.sendInvoiceEmail(
                invoice.getCustomer().getEmail(),
                "Your E-Tour India Invoice – Booking #" + bookingId,
                EmailTemplateUtil.invoiceEmailBody(
                        invoice.getCustomer().getName(),
                        bookingId
                ),
                pdfBytes,
                "ETour-Invoice-" + bookingId + ".pdf"
        );

        return ResponseEntity.ok(
                "Invoice email resend initiated successfully"
        );
    }
    
}

File: dotnet/backend/Controllers/InvoiceController.cs
using Microsoft.AspNetCore.Mvc;
using Etour_Backend_dotnet.Services.Invoice;
using Etour_Backend_dotnet.Services.Email;
using Etour_Backend_dotnet.Utils;

namespace Etour_Backend_dotnet.Controllers;

[ApiController]
[Route("api/invoices")]
public class InvoiceController : ControllerBase
{
    private readonly IInvoiceService _invoiceService;
    private readonly IEmailService _emailService;

    public InvoiceController(IInvoiceService invoiceService, IEmailService emailService)
    {
        _invoiceService = invoiceService;
        _emailService = emailService;
    }

    /// <summary>
    /// Download invoice PDF for a confirmed booking
    /// </summary>
    [HttpGet("{bookingId}/download")]
    public async Task<IActionResult> DownloadInvoice(int bookingId)
    {
        try
        {
            // Generate invoice data (not strictly needed here as PDF gen calls it, but cleaner separation)
            // But GenerateInvoicePdf calls it internally.
            
            // Generate PDF
            byte[] pdfBytes = await _invoiceService.GenerateInvoicePdf(bookingId);

            // Return PDF file
            // Need invoice number for filename. 
            // Optimally we'd get the object first.
            var invoice = await _invoiceService.GenerateInvoice(bookingId);
            
            return File(pdfBytes, "application/pdf", $"{invoice.InvoiceNumber}.pdf");
        }
        catch (Exception ex)
        {
            return BadRequest(new { message = ex.Message });
        }
    }

    /// <summary>
    /// Resend invoice email for a confirmed booking
    /// </summary>
    [HttpPost("{bookingId}/resend-email")]
    public async Task<IActionResult> ResendInvoiceEmail(int bookingId)
    {
        try
        {
            // Generate invoice
            var invoice = await _invoiceService.GenerateInvoice(bookingId);

            // Generate PDF
            byte[] pdfBytes = await _invoiceService.GenerateInvoicePdf(bookingId);

            // Build email body
            string emailBody = $@"
Dear {invoice.Customer.Name},

Thank you for booking with E-Tour India!

Your booking #{bookingId} has been confirmed. Please find your invoice attached to this email.

Booking Details:
- Destination: {invoice.Tour.CategoryName}
- Departure Date: {invoice.Departure.DepartureDate:dd MMM yyyy}
- Duration: {invoice.Departure.NumberOfDays} Days
- Total Amount: ₹{invoice.Amount.TotalAmount:N2}

If you have any questions, please contact us at etour_india@outlook.com.

Happy Travels!
E-Tour India Team
";

            // Send email
            await _emailService.SendInvoiceEmailAsync(
                invoice.Customer.Email,
                $"Your E-Tour India Invoice – Booking #{bookingId}",
                emailBody,
                pdfBytes,
                $"ETour-Invoice-{bookingId}.pdf"
            );

            return Ok(new { message = "Invoice email sent successfully" });
        }
        catch (Exception ex)
        {
            return BadRequest(new { message = ex.Message });
        }
    }
}
